Add sub-RFC for increased availability of NUMA API (#1545)
uxlfoundation/oneTBB: rfcs/proposed/numa_support/increase-numa-support-availability.org

Reviewer: To me, a name better reflecting the gist of the proposal would make more sense. For example, "tbbbind-link-static-hwloc"

Reviewer: Does the title need to get updated as well?

# -*- fill-column: 80; -*-

#+title: Improve predictability of API for NUMA support

*Note:* This is a sub-RFC of the https://github.com/oneapi-src/oneTBB/pull/1535.
Specifically, its section about "Increased availability of NUMA support".

* Introduction
oneTBB has a soft dependency on several variants of ~tbbbind~, which are loaded
by the library as part of its initialization stage. In turn, each ~tbbbind~ has

> a hard dependency, i.e., relies on load-time linking, on a concrete version of

Reviewer: That's OK for me.

the HWLOC library [1, 2]. The soft dependency of oneTBB on ~tbbbind~ allows the
library to continue its execution even if the system loader is unable to resolve
the hard dependency on HWLOC for ~tbbbind~. In this case, the HW topology is not
discovered and the machine is seen as if all CPU cores were uniform, which is
the default TBB behavior when NUMA constraints are not used. Thus, the following

> code returns meaningless values as these values are just ignored by oneTBB:

Reviewer: Would it be better to say that the returned values do not reflect the real topology of the machine? I am not sure in which sense these values are ignored. If you use these to create constrained arenas (well, a single arena), they are not formally ignored, rather they do not really matter.

#+begin_src C++
std::vector<oneapi::tbb::numa_node_id> numa_nodes = oneapi::tbb::info::numa_nodes();
std::vector<oneapi::tbb::core_type_id> core_types = oneapi::tbb::info::core_types();
#+end_src

> An error is also not reported and the client code that uses NUMA support
> facilities may continue running expecting it to work as it was intended. Such
> behavior is not readily noticable by developers that use oneTBB and this
> represents the main problem with the current behavior.

Reviewer: I keep insisting that the main problem is not the "silence", thus the following re-wording:
```suggestion
This lack of valid HW topology data due to absence of a 3rd party library
is the major problem with the current oneTBB behavior.
There is no diagnostics for the issue, which likely makes it unnoticeable
by developers, and the code that uses oneTBB NUMA support
facilities continues running but does not work as it was intended.
```

Having a dependency on a shared HWLOC library has a number of advantages:
1. Code reuse with all of the positive consequences out of this. That's the
   primary purpose of shared libraries.
2. Sharing HWLOC context between its clients. This avoids performing the same
   operations repeatedly with identical results.
3. A drop-in replacement. Users are able to use their own version of HWLOC

>    without recompilation of oneTBB.

Reviewer: My attitude to these "advantages" is that they are of little, if any, importance. No data supporting that "sharing HWLOC context" is provided, so it's kind of unproven. And I simply do not understand why somebody would want to "use their own version of HWLOC" or even what that "special" version would do.

However, importance of these features is not described, neither is a possible impact of their absence, Lack of this information may in general lead to suboptimal decisions.

developers that use oneTBB's NUMA support API need to make sure the library is
available and can be found by oneTBB. Depending on the distribution model of a
developer's code, this is achieved either by:
1. Asking the end user to have necessary version of a dependency pre-installed.
2. Bundling necessary HWLOC version together with other pieces of a product
   release.

However, the requirement to fulfill one of the above steps for the NUMA API to
start paying off may be considered as an incovenience and, what is more
important, it is not always obvious that one of these steps is needed.
Especially, due to silent behavior in case HWLOC library cannot be found in the
environment.

This proposal suggests an improvement to reduce the effect of the disadvantage
being dependent on a dynamic version of HWLOC library by having it linked
statically with one of the ~tbbbind~ libraries that are distributed together
with oneTBB, yet leaving possibility to specify another version of HWLOC library
if users see the need.

[1] [[https://www.open-mpi.org/projects/hwloc/][HWLOC project main page]]

[2] [[https://github.com/open-mpi/hwloc][HWLOC project repository on GitHub]]

* Proposal

> 1. Introduce new variant of the ~tbbbind~ library with the name ~tbbbind_static~

Reviewer: Possibly an open question: Could it make sense to drop support of HWLOC 1.x and use the name `tbbbind` (with no suffixes) for this new statically linked variant? That would also avoid the need to ship one more binary.

Reviewer: I rather disagree about it being unrelated, as it directly addresses one of the shortcomings of the proposal.

> 1. Introduce new variant of the ~tbbbind~ library with the name ~tbbbind_static~
>    which is statically-linked with HWLOC library and distributed along side with
>    the other ~tbbbind~ variants.

Reviewer: ```suggestion
1. Introduce new variant of the ~tbbbind~ library with the name ~tbbbind_static~
   which is linked with a static HWLOC library and distributed along side with
   the other ~tbbbind~ variants.
```

2. Add loading of ~tbbbind_static~ as the last attempt to resolve the dependency
   on functionality provided by ~tbbbind~ layer.
3. Update the oneTBB documentation considering [[https://oneapi-src.github.io/oneTBB/search.html?q=tbb%3A%3Ainfo][these documentation pages]] to
   include steps determining the variant of ~tbbbind~ being used.

** Advantages
The proposed behavior allows having a mechanism for resolving a dependency on
HWLOC library in case it cannot be found in the environment, while still
preferring user-provided version of HWLOC.

As a result, the problematic use of oneTBB API mentioned above should work as
expected, returning enumerated list of actual NUMA nodes and core types on the
system the code is running on, provided that the loaded HWLOC library works on
that system and that an application properly distributes all binaries of oneTBB,
sets the environment so that the necessary variant of ~tbbbind~ library can be
found and loaded.

** Disadvantages
1. There will be one more variation of a ~tbbbind~ binary to ship in oneTBB
   distribution packages.

> 2. Still silent by default behavior in case user failed to setup environment
>    with their own version of HWLOC library correctly. Although, specifying
>    ~TBB_VERSION=1~ envar will help identifying an issue with an environment
>    setup pretty quickly.

Reviewer: ```suggestion
2. By default still no diagnostics if users failed to setup environment
   with their own version of HWLOC library correctly. Although, specifying
   ~TBB_VERSION=1~ envar will help identifying an environment setup issue
   pretty quickly.
```
